### ISSUE TITLE
Crawler optimisation brainstorming

### ISSUE THREAD

**Issue by oliversauter** [WorldBrain/cortex#59]
Hey folks, 

In this issue we want to collect ideas how we want to improve the crawler. 
Just dump everything you think can improve our search results. Then we talk about them together and prioritise. 

The roadmap looks the following:
 1. Launch basic Prototype **(October 15)**
 2. Bug-Fixing and Stability of Crawler, DB, Cortex, Client **(1 Month)**
 3. Improvement phase
  - Search-Engine
  - MVP Features ([as briefly described here](https://docs.google.com/presentation/d/1Pt2NiJwbXq7PNmuzpkqkgfo00fMVP0bxnZgMEakBKWs/edit#slide=id.g16f6f1ade7_5_393))

#### Goal of Improvement
_example: In order for us not to manually recrawl pages, we could use a crawling mgmt software_

#### Technologies we can use
_example: we could use [frontera](frontera.readthedocs.io)_

#### Side notes
_blabalablala_

**Comment by oliversauter** [comment]
From @arpitgogia 

**Goal of Improvement**

Adding Wikipedia to our search results

**Technologies we can use**

We can download the whole wiki database and index it.

**Side notes**

 @dev-tim already did that to test search at the beginning.

**Comment by oliversauter** [comment]
From @arpitgogia 

**Goal of Improvement**

Adding Youtube Videos to our search results
**Technologies we can use**

Via the Youtube API, we can also add search through the transcript of the videos. 

**Side notes**

**Comment by vpolisky** [comment]
**Goal of Improvement**

Spider could/should run fully independently from the cortex app

**Technologies we can use**

Probably, nothing except what are already using

**Side notes**

Spider is actually an independent module and does not really need the cortex app. It connects to the RabbitMQ and to the database. The reason why it's dependent from the cortex app is that spider uses the Django models. To separate spider from cortex, we should find a way how to use these models without the entire cortex. From my view, it's rather a problem of deployment than of implementation. A solution could probably be to make a lightweight Django app which only defines the same models and connects to the same database as cortex.

But does it really make sense?

**Comment by arpitgogia** [comment]
This is an interesting problem. I believe the only way to achieve this is by exposing the database using a set of APIs in the Django app. A big negative with that is the performance reduction due to constant API calls. However since the crawler will then run independently, maybe we can trade off a bit on performance?

**Comment by oliversauter** [comment]
@vpolisky @arpitgogia Yeah I think in the mid run it makes sense to externalise the crawler, since the requests won't influence the cortex performance. 

Could be that if we have a lot of user requests that the crawler cannot run/process data smoothly anymore. 

Correct?

**Comment by vpolisky** [comment]
@oliversauter What do you precisely mean under _user requests_?

The spider runs fully independently from the cortex app. The only intersection of the both is the database. For this, I propose to use a replica set such that writes from the spider don't influence the performance of the database on the user side.

Another influence source is that spider consumes the computer resources. But except that, there should be no ways of influencing.

On the spider side, an asynchronous message consumer - which can/should be run in multiple processes - consumes messages from the message queue and spawns spider processes for each message. It works independently from the Django request-response cycle.

**Comment by oliversauter** [comment]
>It works independently from the Django request-response cycle.

Yeah thats what I was the intention of my question. 

> What do you precisely mean under user requests?

I meant the user request-response cycle and how (if its a lot of search requests processed by cortex) influences the performance of the crawler.

**Comment by vpolisky** [comment]
Ok, understand. So, as already mentioned, the most critical problem, from my view, is the database. This is what will make problems as first. But it is independent of where the spider is actually running. That's why, I think, we need replication.

BTW I assume, that indexing also makes database operations?

**Comment by oliversauter** [comment]
> BTW I assume, that indexing also makes database operations?

@tobisanya this one is for you

> I think, we need replication.

Do you have anything for me to read about that? Or explain it to me. Am I understanding it correctly with that we need multiple DBs to which the requests are sent?

**Comment by tobisanya** [comment]
> BTW I assume, that indexing also makes database operations?
 
For now, indexing only reads from the database, no write is done

**Comment by vpolisky** [comment]
> Do you have anything for me to read about that? Or explain it to me. Am I understanding it correctly with that we need multiple DBs to which the requests are sent?

Yes, this means that we are running multiple database instances on different machines. One instance is used to write into the database. It "copies" its content to other instances. This way, we can separate reads (user requests) from writes (e.g. crawler), such that all write operations are performed on the one instance while user requests are sent to other instances. On the other hand, it improves the availability of the database system, since if one instance goes down, user requests can be automatically redirected to other instances.
